scanner/engine/worker_scheduler.cpp

```diff
@@ -0,0 +1,2724 @@
+          eval_queues[ki];
+      // HACK(apoms): we assume all ops in a kernel group use the
+      //   same number of devices for now.
+      // for (size_t i = 0; i < group.size(); ++i) {
+      KernelFactory* factory = nullptr;
+      for (size_t i = 0; i < group.size(); ++i) {
+        if (std::get<0>(group[i]) != nullptr) {
+          factory = std::get<0>(group[i]) ;
+        }
+      }
+      DeviceType device_type = DeviceType::CPU;
+      i32 max_devices = 1;
+      // Factory should only be null if we only have builtin ops
+      if (factory != nullptr) {
+        device_type = factory->get_device_type();
+        max_devices = factory->get_max_devices();
+        if (max_devices == Kernel::UnlimitedDevices) {
+          max_devices = 1;
+        }
+      }
+      if (device_type == DeviceType::CPU) {
+        for (i32 i = 0; i < max_devices; ++i) {
+          i32 device_id = 0;
+          next_cpu_num++ % num_cpus;
+          for (size_t i = 0; i < group.size(); ++i) {
+            KernelConfig& config = std::get<1>(group[i]);
+            config.devices.clear();
+            config.devices.push_back({device_type, device_id});
+          }
+        }
+      } else {
+        for (i32 i = 0; i < max_devices; ++i) {
+          i32 device_id = gpu_ids[next_gpu_idx++ % num_gpus];
+          for (size_t i = 0; i < group.size(); ++i) {
+            KernelConfig& config = std::get<1>(group[i]);
+            config.devices.clear();
+            config.devices.push_back({device_type, device_id});
+          }
+        }
+      }
+      // Get the device handle for the first kernel in the pipeline
+      if (kg == 0) {
+        if (group.size() == 0) {
+          first_kernel_type = CPU_DEVICE;
+        } else {
+          first_kernel_type = std::get<1>(group[0]).devices[0];
+        }
+      }
+
+      // Input work queue
+      EvalQueue* input_work_queue = &work_queues[kg];
+      // Create new queue for output, reuse previous queue as input
+      EvalQueue* output_work_queue = &work_queues[kg + 1];
+      // Create eval thread for passing data through neural net
+      thread_qs.push_back(
+          std::make_tuple(input_work_queue, output_work_queue));
+      thread_args.emplace_back(EvaluateWorkerArgs{
+          // Uniform arguments
+          node_id_, startup_lock, startup_cv, startup_count,
+
+          // Per worker arguments
+          ki, kg, groups[kg], job_params->boundary_condition(),
+          eval_thread_profilers[kg + 1], results[kg]});
+      eval_total += 1;
+    }
+    // Pre evaluate worker
+    {
+      EvalQueue* input_work_queue;
+      if (distribute_work_dynamically) {
+        input_work_queue = &initial_eval_work[ki];
+      } else {
+        input_work_queue = &initial_eval_work[0];
+      }
+      EvalQueue* output_work_queue =
+          &work_queues[0];
+      assert(groups.size() > 0);
+      pre_eval_queues.push_back(
+          std::make_tuple(input_work_queue, output_work_queue));
+      
```

Comment:
Just fixed it. That's because I erroneously resolved a conflict after rebase. Also rebased to the latest commit. Sorry for making the commit history a little dirty!